onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -240,12 +279,26 @@ Status ModelBuilder::RegisterInitializers() {
     uint32_t index;
     size_t size, padded_size;
     std::tie(index, size, padded_size) = initializers[i++];
-    const char* src = nullptr;
-    if (tensor.data_type() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
-      src = tensor.float_data().empty()
-                ? tensor.raw_data().data()
-                : reinterpret_cast<const char*>(tensor.float_data().data());
-    }  // We should not get anything else here since we already checked in the 1st pass
+    const uint8_t* src = nullptr;
+    // uint8_t data need unpack, need a holder for free memory after copy
+    std::unique_ptr<uint8_t[]> unpacked_tensor;
+    switch (tensor.data_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        src = reinterpret_cast<const uint8_t*>(GetTensorFloatData(tensor));
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_UINT8:
+        size_t tensor_byte_size;
+        ORT_RETURN_IF_ERROR(
+            onnxruntime::utils::UnpackInitializerData(tensor, graph_viewer_.ModelPath(),
+                                                      unpacked_tensor, tensor_byte_size));
+        ORT_RETURN_IF_NOT(size == tensor_byte_size,
+                          "initializer tensor: ", tensor.name(), "'s size: ", tensor_byte_size,
+                          " should match the calculated size: ", size);
+        src = unpacked_tensor.get();
+        break;
+        // default:
```

Comment:
Would it be more consistent to add GetTensorUInt8Data? It should still have the external data check I would have thought.

Is it possible to use UnpackInitializerData for all the types instead of re-implementing some of that in GetTensor*Data? Wondering if a generic helper could check there's no external data and call UnpackInitializerData after that so we don't need another set of type specific helpers.

Reply:
We do check if an initializer used by a node is external in HasExternalInitializer(op_support_checker.cc), so we cannot have any initializer with external data here,

The only issue using UnpackInitializerData  instead of GetTensor*Data is that we need make a copy of the tensor data, especially for the float weights which is still the most popular initializer type.

Reply:
OK. Would be good to have a comment somewhere explaining the reasoning.

---
In reply to: [600162844](https://github.com/microsoft/onnxruntime/pull/7093#discussion_r600162844) [](ancestors = 600162844)

Reply:
Will add comments in a later PR